// prometheus/dist/shared/persistence/persistencia.js
import { promises as fs } from 'node:fs';
import * as fsCb from 'node:fs';
import path from 'node:path';
import { ExcecoesMensagens } from '../../core/messages/core/excecoes-messages.js';
const RAIZ = process.cwd();
const IS_TEST = (process.env.VITEST ?? '') !== '';
function safeGet(obj, key) {
    try {
        return obj[key];
    }
    catch {
        return undefined;
    }
}
function assertInsideRoot(caminho) {
    if ((process.env.VITEST ?? '') !== '' || process.env.PROMETHEUS_ALLOW_OUTSIDE_FS === '1')
        return;
    const resolved = path.resolve(caminho);
    if (!resolved.startsWith(path.resolve(RAIZ))) {
        throw new Error(ExcecoesMensagens.persistenciaNegadaForaRaizProjeto(caminho));
    }
}
function sortKeysDeep(v) {
    if (v === null || v === undefined)
        return v;
    if (Array.isArray(v))
        return v.map(item => sortKeysDeep(item));
    if (typeof v === 'object') {
        const configObject = v;
        const out = {};
        for (const k of Object.keys(configObject).sort())
            out[k] = sortKeysDeep(configObject[k]);
        return out;
    }
    return v;
}
function stableStringify(dados) {
    return JSON.stringify(sortKeysDeep(dados), null, 2);
}
export async function lerEstado(caminho, padrao) {
    try {
        const conteudo = await readFileSafe(caminho, 'utf-8');
        try {
            return JSON.parse(conteudo);
        }
        catch {
            return padrao ?? [];
        }
    }
    catch {
        return padrao ?? [];
    }
}
async function salvarEstadoImpl(caminho, dados) {
    assertInsideRoot(caminho);
    const dir = path.dirname(caminho);
    await mkdirSafe(dir, {
        recursive: true,
        mode: 0o700
    }).catch(() => { });
    const isString = typeof dados === 'string';
    const payload = isString ? dados : stableStringify(dados);
    const tempArquivoCaminho = path.join(dir, `.tmp-bin-${Date.now()}-${Math.random().toString(16).slice(2)}.prometheus`);
    await writeFileSafe(tempArquivoCaminho, payload, {
        encoding: 'utf-8',
        mode: 0o600
    });
    await renameSafe(tempArquivoCaminho, caminho);
}
export let salvarEstado = salvarEstadoImpl;
try {
    const maybeVi = globalThis.vi;
    if (IS_TEST && maybeVi && typeof maybeVi.fn === 'function') {
        salvarEstado = maybeVi.fn(async (...args) => salvarEstadoImpl(...args));
    }
}
catch { }
export async function lerArquivoTexto(caminho) {
    try {
        return await readFileSafe(caminho, 'utf-8');
    }
    catch {
        return '';
    }
}
export async function salvarEstadoAtomico(caminho, dados) {
    assertInsideRoot(caminho);
    const dir = path.dirname(caminho);
    await mkdirSafe(dir, {
        recursive: true,
        mode: 0o700
    });
    const tempArquivoCaminho = path.join(dir, `.tmp-${Date.now()}-${Math.random().toString(16).slice(2)}.json`);
    const payload = stableStringify(dados);
    await writeFileSafe(tempArquivoCaminho, payload, {
        encoding: 'utf-8',
        mode: 0o600
    });
    await renameSafe(tempArquivoCaminho, caminho);
}
export async function salvarBinarioAtomico(caminho, dados) {
    assertInsideRoot(caminho);
    const dir = path.dirname(caminho);
    await mkdirSafe(dir, {
        recursive: true,
        mode: 0o700
    });
    const temporaryValor = path.join(dir, `.tmp-${Date.now()}-${Math.random().toString(16).slice(2)}.bin`);
    const p = fs;
    if (typeof p.writeFile === 'function') {
        await p.writeFile(temporaryValor, dados);
        await renameSafe(temporaryValor, caminho);
        return;
    }
    const cbWrite = safeGet(fsCb, 'writeFile');
    if (typeof cbWrite === 'function') {
        await new Promise((resolve, reject) => {
            cbWrite(temporaryValor, dados, err => err ? reject(err) : resolve());
        });
        await renameSafe(temporaryValor, caminho);
        return;
    }
    if (IS_TEST)
        return;
    throw new Error(ExcecoesMensagens.fsWriteFileBinaryIndisponivel);
}
export let salvarBinario = salvarBinarioAtomico;
try {
    const maybeVi2 = globalThis.vi;
    if (IS_TEST && maybeVi2 && typeof maybeVi2.fn === 'function') {
        salvarBinario = maybeVi2.fn(async (...args) => salvarBinarioAtomico(...args));
    }
}
catch { }
async function readFileSafe(pathname, encoding) {
    const p = fs;
    if (typeof p.readFile === 'function') {
        return await p.readFile(pathname, encoding ?? 'utf-8');
    }
    const cbRead = safeGet(fsCb, 'readFile');
    if (typeof cbRead === 'function') {
        return await new Promise((resolve, reject) => {
            cbRead(pathname, encoding ?? 'utf-8', (err, data) => {
                if (err)
                    reject(err);
                else
                    resolve(data);
            });
        });
    }
    throw new Error(ExcecoesMensagens.fsReadFileIndisponivel);
}
async function writeFileSafe(pathname, data, options) {
    const p = fs;
    if (typeof p.writeFile === 'function') {
        await p.writeFile(pathname, data, options);
        return;
    }
    const cbWrite = safeGet(fsCb, 'writeFile');
    if (typeof cbWrite === 'function') {
        await new Promise((resolve, reject) => {
            cbWrite(pathname, data, options, err => err ? reject(err) : resolve());
        });
        return;
    }
    if (IS_TEST)
        return;
    throw new Error(ExcecoesMensagens.fsWriteFileIndisponivel);
}
async function renameSafe(oldPath, newPath) {
    const p = fs;
    if (typeof p.rename === 'function') {
        await p.rename(oldPath, newPath);
        return;
    }
    const cbRename = safeGet(fsCb, 'rename');
    if (typeof cbRename === 'function') {
        await new Promise((resolve, reject) => {
            cbRename(oldPath, newPath, err => err ? reject(err) : resolve());
        });
        return;
    }
    if (IS_TEST)
        return;
    throw new Error(ExcecoesMensagens.fsRenameIndisponivel);
}
async function mkdirSafe(dirPath, options) {
    const p = fs;
    if (typeof p.mkdir === 'function') {
        await p.mkdir(dirPath, options);
        return;
    }
    const cbMkdir = safeGet(fsCb, 'mkdir');
    if (typeof cbMkdir === 'function') {
        await new Promise((resolve, reject) => {
            cbMkdir(dirPath, options, err => err ? reject(err) : resolve());
        });
        return;
    }
    if (IS_TEST)
        return;
    throw new Error(ExcecoesMensagens.fsMkdirIndisponivel);
}
