// prometheus/dist/core/messages/core/excecoes-messages.js
import { createI18nMessages } from '../../../shared/helpers/i18n.js';
export const ExcecoesMensagens = createI18nMessages({
    exit1: 'exit:1',
    requireMutateFsAutoFix: 'Auto-fix indisponível',
    autoFixTimeout: (timeoutMs) => `Auto-fix timeout após ${timeoutMs}ms`,
    pluginsDesabilitadosSafeMode: 'Carregamento de plugins desabilitado em SAFE_MODE. Defina PROMETHEUS_ALLOW_PLUGINS=1 para permitir.',
    pluginBloqueado: (erro) => `Plugin bloqueado: ${erro}`,
    caminhoPluginNaoResolvido: 'Caminho de plugin não resolvido',
    pluginRegistradoNaoPodeSerObtido: (name) => `Plugin ${name} está registrado mas não pode ser obtido`,
    pluginCarregandoPromiseNaoPodeSerObtida: (name) => `Plugin ${name} está sendo carregado mas promise não pode ser obtida`,
    naoFoiPossivelCarregarPlugin: (name, errMsg) => `Não foi possível carregar o plugin '${name}': ${errMsg}`,
    pluginDeveTerNomeValido: 'Plugin deve ter um nome válido',
    pluginDeveTerVersaoValida: 'Plugin deve ter uma versão válida',
    pluginDeveDefinirPeloMenosUmaExtensao: 'Plugin deve definir pelo menos uma extensão',
    pluginDeveImplementarMetodoParse: 'Plugin deve implementar método parse()',
    definicaoAnalistaInvalida: 'Definição de analista inválida',
    analistaSemFuncaoAplicar: (nome) => `Analista ${nome} sem função aplicar`,
    caminhoForaDaCwdNaoPermitido: (p) => `Caminho fora da CWD não permitido: ${p}`,
    persistenciaNegadaForaRaizProjeto: (caminho) => `Persistência negada: caminho fora da raiz do projeto: ${caminho}`,
    fsWriteFileBinaryIndisponivel: 'fs.writeFile (binary) indisponível no ambiente atual',
    fsReadFileIndisponivel: 'fs.readFile indisponível no ambiente atual',
    fsWriteFileIndisponivel: 'fs.writeFile indisponível no ambiente atual',
    fsRenameIndisponivel: 'fs.rename indisponível no ambiente atual',
    fsMkdirIndisponivel: 'fs.mkdir indisponível no ambiente atual',
    versaoSchemaDesconhecida: (versao) => `Versão de schema desconhecida: ${versao}`,
    relatorioSchemaInvalido: (erros) => `Relatório com schema inválido: ${erros}`,
    arquivoNaoEncontrado: (fileCaminho) => `Arquivo não encontrado: ${fileCaminho}`,
    validacaoFalhouPara: (fileCaminho) => `Validação falhou para ${fileCaminho}`,
    erroAoLer: (fileCaminho, errMsg) => `Erro ao ler ${fileCaminho}: ${errMsg}`,
    erroAoEscrever: (fileCaminho, errMsg) => `Erro ao escrever ${fileCaminho}: ${errMsg}`,
    erroAoDeletar: (fileCaminho, errMsg) => `Erro ao deletar ${fileCaminho}: ${errMsg}`,
    statIndefinidoPara: (fullCaminho) => `Stat indefinido para ${fullCaminho}`,
    mapaReversaoCorrompido: 'Mapa de reversão corrompido',
    semPkg: 'sem pkg'
}, {
    exit1: 'exit:1',
    requireMutateFsAutoFix: 'Auto-fix unavailable',
    autoFixTimeout: (timeoutMs) => `Auto-fix timeout after ${timeoutMs}ms`,
    pluginsDesabilitadosSafeMode: 'Plugin loading disabled in SAFE_MODE. Set PROMETHEUS_ALLOW_PLUGINS=1 to allow.',
    pluginBloqueado: (erro) => `Plugin blocked: ${erro}`,
    caminhoPluginNaoResolvido: 'Plugin path not resolved',
    pluginRegistradoNaoPodeSerObtido: (name) => `Plugin ${name} is registered but cannot be obtained`,
    pluginCarregandoPromiseNaoPodeSerObtida: (name) => `Plugin ${name} is loading but promise cannot be obtained`,
    naoFoiPossivelCarregarPlugin: (name, errMsg) => `Could not load plugin '${name}': ${errMsg}`,
    pluginDeveTerNomeValido: 'Plugin must have a valid name',
    pluginDeveTerVersaoValida: 'Plugin must have a valid version',
    pluginDeveDefinirPeloMenosUmaExtensao: 'Plugin must define at least one extension',
    pluginDeveImplementarMetodoParse: 'Plugin must implement parse() method',
    definicaoAnalistaInvalida: 'Invalid analyst definition',
    analistaSemFuncaoAplicar: (nome) => `Analyst ${nome} without apply function`,
    caminhoForaDaCwdNaoPermitido: (p) => `Path outside CWD not allowed: ${p}`,
    persistenciaNegadaForaRaizProjeto: (caminho) => `Persistence denied: path outside project root: ${caminho}`,
    fsWriteFileBinaryIndisponivel: 'fs.writeFile (binary) unavailable in current environment',
    fsReadFileIndisponivel: 'fs.readFile unavailable in current environment',
    fsWriteFileIndisponivel: 'fs.writeFile unavailable in current environment',
    fsRenameIndisponivel: 'fs.rename unavailable in current environment',
    fsMkdirIndisponivel: 'fs.mkdir unavailable in current environment',
    versaoSchemaDesconhecida: (versao) => `Unknown schema version: ${versao}`,
    relatorioSchemaInvalido: (erros) => `Report with invalid schema: ${erros}`,
    arquivoNaoEncontrado: (fileCaminho) => `File not found: ${fileCaminho}`,
    validacaoFalhouPara: (fileCaminho) => `Validation failed for ${fileCaminho}`,
    erroAoLer: (fileCaminho, errMsg) => `Error reading ${fileCaminho}: ${errMsg}`,
    erroAoEscrever: (fileCaminho, errMsg) => `Error writing ${fileCaminho}: ${errMsg}`,
    erroAoDeletar: (fileCaminho, errMsg) => `Error deleting ${fileCaminho}: ${errMsg}`,
    statIndefinidoPara: (fullCaminho) => `Stat undefined for ${fullCaminho}`,
    mapaReversaoCorrompido: 'Rollback map corrupted',
    semPkg: 'no pkg'
});
