// prometheus/dist/cli/commands/comando-metricas.js
import path from 'node:path';
import { ExitCode, sair } from '../helpers/exit-codes.js';
import { config } from '../../core/config/config.js';
import { formatMs } from '../../core/config/format.js';
import { CliComandoMetricasMensagens } from '../../core/messages/cli/cli-comando-metricas-messages.js';
import { ICONES_DIAGNOSTICO, log, logMetricas } from '../../core/messages/index.js';
import { lerEstado, salvarEstado } from '../../shared/persistence/persistencia.js';
import { Command } from 'commander';
const formatarDuracao = (ms) => formatMs(ms);
function agregados(historico) {
    if (!historico.length)
        return null;
    const total = historico.length;
    const somaAnalise = historico.reduce((acc, h) => acc + (h.tempoAnaliseMs ?? 0), 0);
    const somaParsing = historico.reduce((acc, h) => acc + (h.tempoParsingMs ?? 0), 0);
    const analistasMap = new Map();
    for (const h of historico) {
        if (!h.analistas)
            continue;
        for (const a of h.analistas) {
            const dado = analistasMap.get(a.nome) || {
                totalMs: 0,
                execucoes: 0,
                ocorrencias: 0
            };
            dado.totalMs += a.duracaoMs;
            dado.execucoes += 1;
            dado.ocorrencias += a.ocorrencias;
            analistasMap.set(a.nome, dado);
        }
    }
    const analistasOrdenados = [...analistasMap.entries()].sort((a, b) => b[1].totalMs - a[1].totalMs).slice(0, 5).map(([nome, d]) => ({
        nome,
        totalMs: d.totalMs,
        mediaMs: d.totalMs / d.execucoes,
        execucoes: d.execucoes,
        ocorrencias: d.ocorrencias
    }));
    return {
        totalExecucoes: total,
        mediaAnaliseMs: somaAnalise / total,
        mediaParsingMs: somaParsing / total,
        topAnalistas: analistasOrdenados
    };
}
export function comandoMetricas() {
    return new Command('metricas')
        .description(CliComandoMetricasMensagens.descricao)
        .option('-j, --json', CliComandoMetricasMensagens.opcoes.json)
        .option('-l, --limite <n>', CliComandoMetricasMensagens.opcoes.limite, v => Number(v), 10)
        .option('-e, --export <arquivo>', CliComandoMetricasMensagens.opcoes.export)
        .option('-a, --analistas', CliComandoMetricasMensagens.opcoes.analistas)
        .action(async (opts) => {
        try {
            const caminho = config.ANALISE_METRICAS_HISTORICO_PATH;
            const historico = await lerEstado(caminho).catch(() => []);
            const lista = Array.isArray(historico) ? historico : [];
            const ultimos = opts.limite ? lista.slice(-opts.limite) : lista;
            const agg = agregados(lista) || undefined;
            if (opts.export) {
                const destino = path.isAbsolute(opts.export) ? opts.export : path.join(process.cwd(), opts.export);
                await salvarEstado(destino, {
                    exportadoEm: new Date().toISOString(),
                    total: lista.length,
                    historico: lista
                });
                log.sucesso(CliComandoMetricasMensagens.historicoExportado(destino));
                return;
            }
            if (opts.json) {
                console.log(JSON.stringify({
                    total: lista.length,
                    limite: opts.limite,
                    historico: ultimos,
                    agregados: agg
                }, null, 2));
                return;
            }
            logMetricas.execucoesRegistradas(lista.length);
            if (!lista.length) {
                logMetricas.nenhumHistorico();
                return;
            }
            for (const h of ultimos) {
                const timestampISO = new Date(h.timestamp).toISOString();
                log.info(CliComandoMetricasMensagens.linhaExecucao(timestampISO, h.totalArquivos ?? 0, formatarDuracao(h.tempoAnaliseMs ?? 0), formatarDuracao(h.tempoParsingMs ?? 0), h.cacheAstHits ?? 0, h.cacheAstMiss ?? 0));
            }
            if (opts.analistas && agg) {
                log.info(CliComandoMetricasMensagens.linhaEmBranco);
                log.info(CliComandoMetricasMensagens.tituloTopAnalistas(ICONES_DIAGNOSTICO.info));
                for (const a of agg.topAnalistas) {
                    log.info(CliComandoMetricasMensagens.linhaTopAnalista(a.nome, formatMs(a.totalMs), formatMs(a.mediaMs), a.execucoes, a.ocorrencias));
                }
            }
            if (agg) {
                log.info(CliComandoMetricasMensagens.medias(formatMs(agg.mediaAnaliseMs), formatMs(agg.mediaParsingMs)));
            }
        }
        catch (err) {
            log.erro(CliComandoMetricasMensagens.erroProcessar(err instanceof Error ? err.message : String(err)));
            sair(ExitCode.Failure);
        }
    });
}
