// prometheus/dist/guardian/baseline.js
import path from 'node:path';
import { lerEstado, salvarEstado } from '../shared/persistence/persistencia.js';
import { LINHA_BASE_CAMINHO } from './constantes.js';
export const CURRENT_BASELINE_VERSION = 2;
async function lerBaselineCompleto() {
    try {
        const json = await lerEstado(LINHA_BASE_CAMINHO);
        if (json && typeof json === 'object' && !Array.isArray(json)) {
            const base = json;
            if (base.snapshot && typeof base.snapshot === 'object') {
                return base;
            }
            const keys = Object.keys(json);
            if (keys.length > 0 && keys[0]?.startsWith('.')) {
                const entries = Object.entries(json).filter(([k, v]) => typeof k === 'string' && typeof v === 'string');
                return {
                    snapshot: Object.fromEntries(entries),
                    version: 1
                };
            }
            return null;
        }
        return null;
    }
    catch {
        return null;
    }
}
export async function carregarBaseline() {
    const completo = await lerBaselineCompleto();
    return completo?.snapshot ?? null;
}
export async function carregarAssinaturaBaseline() {
    const completo = await lerBaselineCompleto();
    return completo?.assinatura;
}
export async function salvarBaseline(snapshot, assinatura) {
    const fs = await import('node:fs');
    await fs.promises.mkdir(path.dirname(LINHA_BASE_CAMINHO), {
        recursive: true
    });
    const baselineCompleto = {
        snapshot,
        assinatura,
        version: CURRENT_BASELINE_VERSION
    };
    await salvarEstado(LINHA_BASE_CAMINHO, baselineCompleto);
}
