// prometheus/dist/guardian/sentinela.js
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { config } from '../core/config/config.js';
import { log } from '../core/messages/index.js';
import micromatch from 'micromatch';
import { GuardianError, IntegridadeStatus } from '../types/index.js';
import { carregarAssinaturaBaseline, carregarBaseline, salvarBaseline } from './baseline.js';
import { LINHA_BASE_CAMINHO } from './constantes.js';
import { diffSnapshots, verificarErros } from './diff.js';
import { assinarConteudo, verificarAssinatura, verificarGpgInstalado } from './gpg.js';
import { gerarSnapshotDoConteudo } from './hash.js';
function construirSnapshot(fileEntries) {
    const snapshot = {};
    for (const { relPath, content } of fileEntries) {
        if (typeof content !== 'string' || !content.trim())
            continue;
        try {
            snapshot[relPath] = gerarSnapshotDoConteudo(content);
        }
        catch (err) {
            log.aviso(`x Falha ao gerar hash de ${relPath}: ${typeof err === 'object' && err && 'message' in err ? err.message : String(err)}`);
        }
    }
    return snapshot;
}
export async function scanSystemIntegrity(fileEntries, options) {
    const agora = new Date().toISOString();
    await fs.mkdir(path.dirname(LINHA_BASE_CAMINHO), {
        recursive: true
    });
    let baselineAnterior = null;
    try {
        baselineAnterior = await carregarBaseline();
    }
    catch (err) {
        log.aviso(`⚠️ Baseline inválido ou corrompido: ${typeof err === 'object' && err && 'message' in err ? err.message : String(err)}`);
    }
    const ignorados = config.INCLUDE_EXCLUDE_RULES && config.INCLUDE_EXCLUDE_RULES.globalExcludeGlob || [];
    const filtrados = fileEntries.filter(f => {
        const rel = f.relPath.replace(/\\/g, '/');
        return !micromatch.isMatch(rel, ignorados);
    });
    if (config.DEV_MODE) {
        const removidos = fileEntries.length - filtrados.length;
        log.info(`⚙️ Guardian filtro aplicado: ${filtrados.length} arquivos considerados (removidos ${removidos}).`);
    }
    const { gerarSnapshotDoConteudo: gerar } = await import('./hash.js');
    const snapshotAtual = {};
    for (const { relPath, content } of filtrados) {
        if (typeof content !== 'string' || !content.trim())
            continue;
        try {
            snapshotAtual[relPath] = gerar(content);
        }
        catch (err) {
            log.aviso(`\u001Fx Falha ao gerar hash de ${relPath}: ${typeof err === 'object' && err && 'message' in err ? err.message : String(err)}`);
        }
    }
    if (!baselineAnterior) {
        if (options?.justDiff) {
            return {
                status: IntegridadeStatus.Ok,
                timestamp: agora,
                detalhes: []
            };
        }
        const gpgHabilitado = config.GUARDIAN_GPG_ENABLED;
        if (gpgHabilitado && config.GUARDIAN_GPG_KEY_ID) {
            const gpgInstalado = await verificarGpgInstalado();
            if (gpgInstalado && !options?.suppressLogs) {
                log.info(`🛡️ Guardian: GPG habilitado, assinatura será aplicada.`);
            }
        }
        if (!options?.suppressLogs) {
            log.info(`🆕 Guardian: baseline inicial criado.`);
        }
        const jsonString = JSON.stringify(snapshotAtual);
        const assinatura = await assinarConteudo(jsonString);
        await salvarBaseline(snapshotAtual, assinatura ?? undefined);
        return {
            status: IntegridadeStatus.Criado,
            timestamp: agora
        };
    }
    if (process.argv.includes('--aceitar')) {
        if (!options?.suppressLogs) {
            log.info(`✅ Guardian: baseline aceito manualmente (--aceitar).`);
        }
        const jsonString = JSON.stringify(snapshotAtual);
        const assinatura = await assinarConteudo(jsonString);
        await salvarBaseline(snapshotAtual, assinatura ?? undefined);
        return {
            status: IntegridadeStatus.Aceito,
            timestamp: agora
        };
    }
    const gpgHabilitado = config.GUARDIAN_GPG_ENABLED;
    if (gpgHabilitado && !options?.suppressLogs) {
        const assinaturaSalva = await carregarAssinaturaBaseline();
        if (assinaturaSalva) {
            const baselineJson = JSON.stringify(baselineAnterior);
            const verificacao = await verificarAssinatura(baselineJson, assinaturaSalva.assinatura);
            if (verificacao.valido) {
                log.info(`🛡️ Guardian: assinatura GPG verificada - ${verificacao.assinante || 'OK'}`);
            }
            else {
                log.aviso(`🛡️ Guardian: ⚠️ assinatura GPG INVÁLIDA - ${verificacao.mensagem}`);
                if (!options?.justDiff) {
                    const errorDetails = [{
                            tipo: 'assinatura-invalida',
                            mensagem: `Assinatura GPG inválida: ${verificacao.mensagem}`
                        }];
                    throw new GuardianError(errorDetails);
                }
            }
        }
        else if (!config.GUARDIAN_GPG_KEY_ID) {
            log.aviso(`🛡️ Guardian: GPG habilitado mas sem chave configurada, usando apenas hash`);
        }
    }
    const diffs = diffSnapshots(baselineAnterior, snapshotAtual);
    const erros = verificarErros(diffs);
    if (options?.justDiff) {
        return {
            status: erros.length ? IntegridadeStatus.AlteracoesDetectadas : IntegridadeStatus.Ok,
            timestamp: agora,
            detalhes: erros
        };
    }
    if (erros.length) {
        const errorDetails = erros.map(erro => ({
            tipo: 'integridade',
            mensagem: erro
        }));
        throw new GuardianError(errorDetails);
    }
    return {
        status: IntegridadeStatus.Ok,
        timestamp: agora
    };
}
export async function acceptNewBaseline(fileEntries) {
    const ignorados = [];
    const ignoradosDyn = config.INCLUDE_EXCLUDE_RULES && config.INCLUDE_EXCLUDE_RULES.globalExcludeGlob || ignorados;
    const filtrados = fileEntries.filter(f => {
        const rel = f.relPath.replace(/\\/g, '/');
        return !micromatch.isMatch(rel, ignoradosDyn);
    });
    const snapshotAtual = construirSnapshot(filtrados);
    await fs.mkdir(path.dirname(LINHA_BASE_CAMINHO), {
        recursive: true
    });
    const jsonString = JSON.stringify(snapshotAtual);
    const assinatura = await assinarConteudo(jsonString);
    await salvarBaseline(snapshotAtual, assinatura ?? undefined);
}
