// prometheus/dist/shared/helpers/helpers-analistas.js
import { criarOcorrencia } from '../../types/index.js';
export function incrementar(contador, chave) {
    contador[chave] = (contador[chave] ?? 0) + 1;
}
export function garantirArray(valor) {
    return Array.isArray(valor) ? valor : [];
}
export function criarOcorrenciaErroAnalista(info) {
    const mensagem = info.erro instanceof Error
        ? info.erro.message
        : String(info.erro);
    return criarOcorrencia({
        tipo: 'erro-analista',
        nivel: 'aviso',
        mensagem: `Erro no analista ${info.nome}: ${mensagem}`,
        relPath: info.relPath,
        linha: info.linha ?? 1,
        origem: info.nome
    });
}
export function executarComTratamentoErro(fn, nomeAnalista, relPath, linhaErro = 1) {
    try {
        return fn();
    }
    catch (erro) {
        return [criarOcorrenciaErroAnalista({
                nome: nomeAnalista,
                relPath,
                erro,
                linha: linhaErro
            })];
    }
}
export async function executarAsyncComTratamentoErro(fn, nomeAnalista, relPath, linhaErro = 1) {
    try {
        return await fn();
    }
    catch (erro) {
        return [criarOcorrenciaErroAnalista({
                nome: nomeAnalista,
                relPath,
                erro,
                linha: linhaErro
            })];
    }
}
