// prometheus/dist/core/messages/cli/cli-comando-metricas-messages.js
import { createI18nMessages } from '../../../shared/helpers/i18n.js';
export const CliComandoMetricasMensagens = createI18nMessages({
    descricao: 'Inspeciona histórico de métricas de execuções anteriores',
    opcoes: {
        json: 'Saída em JSON estruturado (para CI/integrações)',
        limite: 'Quantidade de registros mais recentes (default 10)',
        export: 'Exporta histórico completo em JSON para arquivo',
        analistas: 'Exibe tabela agregada por analista (top 5)'
    },
    erroProcessar: (erro) => `Falha ao processar métricas: ${erro}`,
    historicoExportado: (destino) => `?? Histórico completo exportado para ${destino}`,
    linhaExecucao: (timestampISO, totalArquivos, duracaoAnalise, duracaoParsing, cacheHits, cacheMiss) => `- ${timestampISO} | arquivos=${totalArquivos} analise=${duracaoAnalise} parsing=${duracaoParsing} cache(h/m)=${cacheHits}/${cacheMiss}`,
    tituloTopAnalistas: (iconeInfo) => `${iconeInfo} Top analistas (por tempo acumulado):`,
    linhaTopAnalista: (nome, total, media, execucoes, ocorrencias) => `  • ${nome} total=${total} média=${media} exec=${execucoes} ocorr=${ocorrencias}`,
    medias: (mediaAnalise, mediaParsing) => `\nMédias (histórico completo): analise=${mediaAnalise}, parsing=${mediaParsing}`,
    linhaEmBranco: ''
}, {
    descricao: 'Inspects metrics history from previous executions',
    opcoes: {
        json: 'Output in structured JSON (for CI/integrations)',
        limite: 'Number of most recent records (default 10)',
        export: 'Exports complete history in JSON to file',
        analistas: 'Displays aggregated table by analyst (top 5)'
    },
    erroProcessar: (erro) => `Failed to process metrics: ${erro}`,
    historicoExportado: (destino) => `?? Complete history exported to ${destino}`,
    linhaExecucao: (timestampISO, totalArquivos, duracaoAnalise, duracaoParsing, cacheHits, cacheMiss) => `- ${timestampISO} | files=${totalArquivos} analysis=${duracaoAnalise} parsing=${duracaoParsing} cache(h/m)=${cacheHits}/${cacheMiss}`,
    tituloTopAnalistas: (iconeInfo) => `${iconeInfo} Top analysts (by accumulated time):`,
    linhaTopAnalista: (nome, total, media, execucoes, ocorrencias) => `  • ${nome} total=${total} avg=${media} exec=${execucoes} occur=${ocorrencias}`,
    medias: (mediaAnalise, mediaParsing) => `\nAverages (complete history): analysis=${mediaAnalise}, parsing=${mediaParsing}`,
    linhaEmBranco: ''
});
