// prometheus/dist/core/messages/cli/cli-exibir-moldura-messages.js
import { createI18nMessages } from '../../../shared/helpers/i18n.js';
export const CliExibirMolduraMensagens = createI18nMessages({
    fallbackLinha: (linha) => `  ${linha}`,
    planoTitulo: 'Plano de reestruturação',
    planoCabecalhoLinha1: 'De                                → Para',
    planoCabecalhoLinha2: '----------------------------------  ---------------------------------------',
    planoOverflow: (restantes) => `... +${restantes} restantes`,
    planoFallbackLinha: (de, para) => `  - ${de} → ${para}`,
    planoFallbackOverflow: (restantes) => `  ... +${restantes} restantes`,
    conflitosTitulo: 'Conflitos de destino',
    conflitosCabecalhoLinha1: 'Destino                           Motivo',
    conflitosCabecalhoLinha2: '-------------------------------   ------------------------------',
    conflitosOverflow: (restantes) => `... +${restantes} restantes`,
    conflitosFallbackLinha: (alvo, motivo) => `  - ${alvo} :: ${motivo}`,
    conflitosFallbackOverflow: (restantes) => `  ... +${restantes} restantes`
}, {
    fallbackLinha: (linha) => `  ${linha}`,
    planoTitulo: 'Restructuring plan',
    planoCabecalhoLinha1: 'From                              → To',
    planoCabecalhoLinha2: '----------------------------------  ---------------------------------------',
    planoOverflow: (restantes) => `... +${restantes} remaining`,
    planoFallbackLinha: (de, para) => `  - ${de} → ${para}`,
    planoFallbackOverflow: (restantes) => `  ... +${restantes} remaining`,
    conflitosTitulo: 'Destination conflicts',
    conflitosCabecalhoLinha1: 'Destination                      Reason',
    conflitosCabecalhoLinha2: '-------------------------------   ------------------------------',
    conflitosOverflow: (restantes) => `... +${restantes} remaining`,
    conflitosFallbackLinha: (alvo, motivo) => `  - ${alvo} :: ${motivo}`,
    conflitosFallbackOverflow: (restantes) => `  ... +${restantes} remaining`
});
