// prometheus/dist/core/messages/cli/cli-comando-analistas-messages.js
import { createI18nMessages } from '../../../shared/helpers/i18n.js';
export const CliComandoAnalistasMensagens = createI18nMessages({
    descricao: 'Lista analistas registrados e seus metadados atuais',
    opcoes: {
        json: 'Saída em JSON estruturado (para CI/integrações)',
        output: 'Arquivo para exportar JSON de analistas',
        doc: 'Gera documentação Markdown dos analistas'
    },
    erroListar: (erro) => `Falha ao listar analistas: ${erro}`,
    fastModeTitulo: '\n📋 Analistas registrados (FAST MODE):\n',
    fastModeTotalZero: '\nTotal: 0',
    docMdTitulo: 'Analysts Documentation',
    docGeradoEm: (iso) => `Gerado em: ${iso}`,
    docTabelaHeader: '| Nome | Categoria | Descrição | Limites |',
    docTabelaSeparador: '| ---- | --------- | --------- | ------- |',
    docLinhaAnalista: (nome, categoria, descricao, limitesStr) => `| ${nome} | ${categoria} | ${descricao} | ${limitesStr} |`,
    docGerada: (destinoDoc) => `📄 Documentação de analistas gerada em ${destinoDoc}`,
    jsonExportado: (destino) => `📄 Exportado JSON de analistas para ${destino}`,
    titulo: '\n📋 Analistas registrados:\n',
    linhaAnalista: (nome, categoria, descricao) => `- ${nome} (${categoria}) ${descricao ? `: ${descricao}` : ''}`,
    tituloComIcone: (iconeInfo) => `${iconeInfo} Analistas registrados:`,
    total: (n) => `\nTotal: ${n}`
}, {
    descricao: 'Lists registered analysts and their current metadata',
    opcoes: {
        json: 'Output in structured JSON (for CI/integrations)',
        output: 'File to export analysts JSON',
        doc: 'Generates Markdown documentation of analysts'
    },
    erroListar: (erro) => `Failed to list analysts: ${erro}`,
    fastModeTitulo: '\n📋 Registered analysts (FAST MODE):\n',
    fastModeTotalZero: '\nTotal: 0',
    docMdTitulo: 'Analysts Documentation',
    docGeradoEm: (iso) => `Generated at: ${iso}`,
    docTabelaHeader: '| Name | Category | Description | Limits |',
    docTabelaSeparador: '| ---- | --------- | --------- | ------- |',
    docLinhaAnalista: (nome, categoria, descricao, limitesStr) => `| ${nome} | ${categoria} | ${descricao} | ${limitesStr} |`,
    docGerada: (destinoDoc) => `📄 Analyst documentation generated at ${destinoDoc}`,
    jsonExportado: (destino) => `📄 Exported analysts JSON to ${destino}`,
    titulo: '\n📋 Registered analysts:\n',
    linhaAnalista: (nome, categoria, descricao) => `- ${nome} (${categoria}) ${descricao ? `: ${descricao}` : ''}`,
    tituloComIcone: (iconeInfo) => `${iconeInfo} Registered analysts:`,
    total: (n) => `\nTotal: ${n}`
});
