---
Proveni√™ncia e Autoria: Este documento integra o projeto Prometheus (licen√ßa MIT-0).
Nada aqui implica cess√£o de direitos morais/autorais.
Conte√∫dos de terceiros n√£o licenciados de forma compat√≠vel n√£o devem ser inclu√≠dos.
Refer√™ncias a materiais externos devem ser linkadas e reescritas com palavras pr√≥prias.
---


# import/no-empty-named-blocks

üîßüí° This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix) and manually fixable by [editor suggestions](https://eslint.org/docs/latest/use/core-concepts#rule-suggestions).

<!-- end auto-generated rule header -->

Reports the use of empty named import blocks.

## Rule Details

### Valid

```js
import { mod } from 'mod'
import Default, { mod } from 'mod'
```

When using typescript

```js
import type { mod } from 'mod'
```

When using flow

```js
import typeof { mod } from 'mod'
```

### Invalid

```js
import {} from 'mod'
import Default, {} from 'mod'
```

When using typescript

```js
import type Default, {} from 'mod'
import type {} from 'mod'
```

When using flow

```js
import typeof {} from 'mod'
import typeof Default, {} from 'mod'
```

